### ISSUE TITLE
Using Prisma with a PlanetScale database

### ISSUE THREAD

**Issue by janpio** [prisma/prisma#7292]
[PlanetScale](https://planetscale.com) is a new and interesting cloud database based on MySQL. It's exciting!

PlanetScale is a bit different than other MySQL database providers though:

1. **No support for foreign keys**. This is challenging for Prisma users, as Prisma by default uses foreign keys to express relations.
1. **No creation of additional databases using `CREATE DATABASE`** but need to use their tooling (web UI or CLI). This can be challenging, as some development commands of Prisma like `prisma migrate dev` create temporary databases in the background.
1. **No schema changes on your production database** (but make them on branches, then merge the schema changes back). This is challenging for Prisma users, as usually they run Migrations against their database directly via `prisma migrate deploy` or `prisma db push`.

---

Here is how you can use Prisma and PlanetScale together anyway:
 
1. **No foreign keys** / No Prisma Migrate (`migrate dev` and `db push`)
	- Problem: PlanetScale does not allow foreign keys in the database schema (`ERROR 70100 (1317): foreign key constraints are not allowed, see https://code.openark.org/blog/mysql/the-problem-with-mysql-foreign-key-constraints-in-online-schema-changes`), which Prisma relies on for relations. 
	- Starting with 2.27.0 Prisma catches the error and outputs a helpful error message (that currently redirects here)	
	- Workaround starting with 3.1.1: Use the preview feature `referentialIntegrity` and the datasource propertly `referentialIntegrity = "prisma"` to enable a mode that automatically leaves out foreign keys in migrations:
	  ```
	  generator client {
	    provider        = "prisma-client-js"
	    previewFeatures = ["referentialIntegrity"]
	  }
	  
	  datasource db {
	    provider             = "mysql"
	    url                  = env("DATABASE_URL")
	    shadowDatabaseUrl    = env("SHADOW_DATABASE_URL")
	    referentialIntegrity = "prisma"
	  }
	  
	  ...
	  ```
	  You can now properly use `prisma migrate dev` and `prisma db push` ✅
	  (Between 2.24.0 and 3.1.1 the setting was called `planetScaleMode = true` and hidden behind the preview feature flag `planetScaleMode`)
	

2. If you are using `migrate dev` to migrate: **No `CREATE DATABASE` / No automatic shadow database creation**
	- Problem: PlanetScale does not allow creating new databases with `CREATE DATABASE`, which Prisma Migrate prefers to use for the shadow database of Prisma Migrate. 
	- Solution: Create a branch `shadow` or similar and open put its connection string as `shadowDatabase` of your `datasource` in `schema.prisma` ✅
	  ```
	  datasource db {
	    provider             = "mysql"
	    url                  = env("DATABASE_URL")
	    shadowDatabaseUrl    = env("SHADOW_DATABASE_URL")
	    referentialIntegrity = "prisma"
	  }
	  ```
	- Potential improvement: Catch error thrown if you try anyway and output better error message with link to documentation.

3. If you are using `migrate dev` to migrate: **Prisma migration table is not copied to `main` by default**
   - PlanetScale now has a setting "Automatic schema migrations" that enables this behavior:
	![image](https://user-images.githubusercontent.com/183673/127225144-cfc757c3-ee73-4b97-afbd-5650b3aabe3f.png)
    As you can see `Prisma` is one of the default options, and enabling this option and choosing `Prisma` makes sure that the content of the migration table `_prisma_migrations` is copied from the branch to `main` with the deploy request. 🥳

4. **No schema changes on production branches**
    - Problem: PlanetScale does not allow schema changes on the production branch of the database (`ERROR HY000 (1105): direct DDL is disabled`), which Prisma Migrate tries to do if you tell it to.
    - Starting with 2.27.0 Prisma catches the error and outputs a helpful error message (that currently redirects here)
	- Solution: Only execute Prisma Migrate on non production branches, then use a PlanetScale "deploy request" to merge the schema changes to your `main` ✅

We are very excited for you to try PlanetScale with Prisma. If you hit any bumps, let us know either here or in new issues or discussions in our repository. Thanks!

**Comment by heymartinadams** [comment]
Holy sh*t @janpio — that’s awesome of you to put this together! Thank you! 💚

We’ll stick with AWS for now until PlanetScale is a bit more out of beta and has enabled serverless deployment connections.

**Comment by janpio** [comment]
I updated the original post:
- More concrete instructions on how do work get the shadow database working (which works great!)
- Added a "Potential improvement" point to all problems, to show and communicate what we are thinking about right now

**Comment by janpio** [comment]
I updated the original post:

- Improved formatting and added error message for google-ability
- Added links to the two "Potential improvements" we have planned on our side
- Mentioned our brand new workaround `planetScaleMode = true` for the "No foreign keys" problem in Prisma Migrate (added in 2.24.0)

**Comment by bigint** [comment]
Planetscal serverless is available now 🎉

https://vercel.com/integrations/planetscale

**Comment by janpio** [comment]
I updated the original post:

- Link to https://github.com/planetscale/beta/discussions/15 for the "No connection URL" and "No serverless connection" secttions as PlanetScale has communicated they are working on connection string support which will solve both these problems ⛰️ 
 - Mention for the both limitations errors (direct DDL is disabled, foreign key constraints are not alloweed) that Prisma will provide a helpful error message about these starting in 2.27.0.
 - Mention for "No support for Prisma migration table" that PlanetScale and Prisma are finishing implementation of support for this.

**Comment by janpio** [comment]
I just updated the post:

- with the brand new `Automatic schema migrations` feature PlanetScale launched today, that includes support for Prisma's `_prisma_migrations` table that holds the migration meta information. With that it is now much easier to use Prisma Migrate on branches, and then use PlanetScale's deploy requests to merge the schema changes back to the `main` branch. Nice - thanks @planetscale 💚

**Comment by janpio** [comment]
I just updated the post:

- Mention the use of `--execute-protocol 'mysql'` in both 1) and 4) to make the `pscale` CLI to generate an env var with the `mysql://` protocol vs. the default `mysql2://` one that Prisma does not support.
  - Pssst, you can also add e.g. `--execute-env-url 'PLANETSCALE_DATABASE_URL'` if you want it to put the value into another env var than `DATABASE_URL`.

**Comment by heymartinadams** [comment]
> Planetscale serverless is available now 🎉
> 
> https://vercel.com/integrations/planetscale

@bigint Following these directions, using the Vercel Integration, I was able to connect to PS on localhost by providing the issued `127.0.0.1:3306` string into a `DATABASE_URL` secret, but as soon as I shifted to deploy it on Vercel, using *as part of a build script* the following command, no connection string was printed or issued:

`pscale connect my_database_name branch_name --execute-protocol 'mysql' --service-token-name ${process.env.PLANETSCALE_TOKEN_NAME} --service-token ${process.env.PLANETSCALE_TOKEN}  --execute 'next build'`

Do you know if that’s still being developed, @janpio?

**Comment by janpio** [comment]
As far as I am aware the PlanetScale Vercel integration, and the CLI are two different things. The integration creates databases for preview deploys and so on - so it is very possible that for example the tokens do not have access to your other databases for example.

I would wait until PlanetScale offers a standard connection string before continuing to play with Vercel. The CLI is just not a good match for that right now unfortunately.

**Comment by bigint** [comment]
This is finally available 🎉 

https://twitter.com/planetscaledata/status/1427371097352069120

![image](https://user-images.githubusercontent.com/69431456/129652562-1c79b91b-4c48-440a-9c40-51f4f03e2c4d.png)

**Comment by janpio** [comment]
Testing the details of this right now and will update the message above soon. Looking good!

**Comment by heymartinadams** [comment]
Great, @janpio! Been working on it since yesterday — and was able to successfully connect to the database on Vercel using their connection string.

The issue we’re running into now is that deploys into PlanetScale’s `main` database aren’t working (which is handled via their CLI or their dashboard).

<img width="794" alt="CleanShot 2021-08-17 at 10 46 50@2x" src="https://user-images.githubusercontent.com/11673745/129775587-7eac7e74-0287-484a-9d38-48fd0788e1a7.png">

**Comment by bigint** [comment]
`onDelete: cascade` is not working :(

**Comment by heymartinadams** [comment]
@bigint yeah 😕 but man, have you seen the speed?! Incredible.

**Comment by thebiglabasky** [comment]
Indeed, referential actions aren't supported, because PlanetScale does not support foreign keys, so there isn't cascading behavior out of the box.

**Comment by bigint** [comment]
Are foreign keys bad in general for scalable apps? 🤔

**Comment by pantharshit00** [comment]
Read this classic response from GitHub's Ghost team on this: https://github.com/github/gh-ost/issues/331#issuecomment-266027731

Basically foreign keys makes sharding hard and sharding is a core part of planetscale which is based on the vitess sharding manager.

**Comment by fatih** [comment]
One of our engineers at PlanetScale wrote an extensive blog post about it if you're interested into it: http://code.openark.org/blog/mysql/the-problem-with-mysql-foreign-key-constraints-in-online-schema-changes

**Comment by bigint** [comment]
So we need to write the cascade delete query manually? by finding and deleting them?

**Comment by janpio** [comment]
@heymartinadams Hmm, that is new - or we somehow never triggered it. The only tables in Prisma that do not have a primary key should be the implicit m:n relation tables (Can you see in the schema diff which one it is complaining about?). Maybe ask about this limitation over at https://github.com/planetscale/beta/discussions as PlanetScale probably can tell you much quicker what is going on.

**Comment by heymartinadams** [comment]
@janpio I’m in touch with @nickvanw of PlanetScale to resolve the issue. Indeed, the tables in question that currently prevent us from using PlanetScale’s database branch deploys to the `main` database are all `m:n` relation tables. Nick is looking into whether or not PS will keep that requirement, or if they can just use Prisma’s `UNIQUE INDEX` for those tables.

**Comment by nickvanw** [comment]
Thanks for looping me in @heymartinadams, I was actually just going to write you back! 

We took a look at that example table that Prisma made this morning, and determined that the `UNIQUE INDEX` that Prisma creates should be enough for the system to maintain the proper integrity, and slightly relaxed the requirement for an explicit `PRIMARY KEY` that we had previously.

This change was deployed this morning, and I've tested personally that a table looking like this:

<img width="662" alt="Screen Shot 2021-08-18 at 3 46 56 PM" src="https://user-images.githubusercontent.com/1907071/129981845-3989f399-8714-4c63-859e-c658ad3d1107.png">

Can successfully be created and merged using a Deploy Request.

This should mean that existing schema generated by Prisma should be able to work with no modification, and you shouldn't need to modify Prisma's migration files moving forward.

As always, let us know if there's anything you run into that doesn't look right. We're having a blast jumping through these issues and making PlanetScale a better platform for everyone - thank you so much for spending the time working with us!

**Comment by janpio** [comment]
Nice outcome, seems Prisma and PlanetScale have similar requirements about being able to uniquely identify each row in a table either by PK or Unique index 👍

**Comment by janpio** [comment]
I just updated the post:
- Removed all outdated and striked through information to make it easier to read and prepare a big rewrite.

**Comment by heymartinadams** [comment]
Thank you, @janpio. We’ve been testing PlanetScale and just started using it in production since yesterday, and it’s blazingly fast. Together with Prisma our backend’s becoming a powerhouse. 🚀 

Doesn’t your point #6 above need to be crossed out/removed? Asking because we were able to deploy PlanetScale without a hitch to our serverless environment (using Vercel). https://planetscale.com/blog/connect-any-mysql-client-to-planetscale-using-connection-strings

**Comment by jferrettiboke** [comment]
I have been exactly following the [Automatic Prisma Migrations](https://docs.planetscale.com/tutorials/automatic-prisma-migrations) tutorial, and it went all good until I needed to change my schema after the first migration. I don't know if this happened to you guys.

I'm getting this error in my terminal when trying to create more users with `phoneNumber`.

```
Unknown arg `phoneNumber` in data.phoneNumber for type UserCreateInput. Did you mean `name`? Available args:
type UserCreateInput {
  email: String
  name?: String | Null
  posts?: PostCreateNestedManyWithoutAuthorInput
  profile?: ProfileCreateNestedOneWithoutUserInput
}
```

`UserCreateInput` in `node_modules/.prisma/client/index.d.ts`:

```ts
  export type UserCreateInput = {
    email: string
    name?: string | null
    phoneNumber?: string | null
    posts?: PostCreateNestedManyWithoutAuthorInput
    profile?: ProfileCreateNestedOneWithoutUserInput
  }
```

**Comment by thebiglabasky** [comment]
Looks like the client might not have been regenerated?

**Comment by jferrettiboke** [comment]
@thebiglabasky According to the type, it has been generated. Everything seems to be fine in the PlanetScale database though (`phoneNumber` is equals to `null`), but Prisma is not able to store or retrieve this field. Using `select: { phoneNumber: true }` doesn't work either. As I said, this happens after the second migration. First migration went all good.

**Comment by janpio** [comment]
> Doesn’t your point 6. above need to be crossed out/removed? Asking because we were able to deploy PlanetScale without a hitch to our serverless environment (using Vercel). [planetscale.com/blog/connect-any-mysql-client-to-planetscale-using-connection-strings](https://planetscale.com/blog/connect-any-mysql-client-to-planetscale-using-connection-strings)

Yes (also point 1), the general structure of the post does not really make sense any more so we will rewrite and restructure it.

**Comment by thebiglabasky** [comment]
@jferrettiboke Hum, that looks like deserving opening a separate issue with details for us to replicate it and fix 🙏

**Comment by janpio** [comment]
Yes, they totally can be used together. The opinionations or limitations that PlanetScale has in comparison to plain MySQL (no FKs, no DDL on production branch, no manual database creation) require you to work a certain way and possibly adapt your workflow but generally it does work (e.g. migrations can be done this way: https://docs.planetscale.com/tutorials/automatic-prisma-migrations). 

(I'll try to update the original issue above later today, swamped with other stuff unfortunately)

**Comment by 2color** [comment]
As mentioned in the issue, when using Prisma with PlanetScale, you can't define foreign keys and rely on cascading deletes.

This means that the responsibility of managing cascading deletes(to avoid dangling relations on deletes) and other [referential actions](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/referential-actions) falls on you as a developer.

PlanetScale has recently released a [guide on operating without foreign keys](https://docs.planetscale.com/tutorials/operating-without-foreign-keys), which I highly recommend to developers using Prisma with PlanetScale.

**Comment by harshamv** [comment]
@2color thanks. is there any guide to understanding what changes we need to do in our Prisma schema to get this to work?

**Comment by tomhoule** [comment]
There is a [new preview feature in 3.1.1](https://github.com/prisma/prisma/releases/tag/3.1.1) made with PlanetScale support in mind: `referentialIntegrity`.

Concretely, it lets you write a schema with relations, where the relations are not backed by foreign keys but Prisma emulates referential actions. That will also prevent Migrate from adding foreign keys to your tables. With that, you should be able to use Prisma (nearly) the same as on a regular MySQL installation.

It's used like in this example:

```
datasource db {
  provider             = "mysql"
  url                  = env("DATABASE_URL")
  referentialIntegrity = "prisma"
}

generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["referentialIntegrity"]
}
```

And there is a [docs page](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/referential-integrity) about it.

Feedback of any form is very much appreciated :)

**Comment by janpio** [comment]
I just updated the post:

- Removed limitations around connection strings and serverless - as that is now provided by PlanetScale out of the box 💕 
- Updated our configuration from `planetScaleMode` to `referentialIntegrity`
- Reordered remaining list items by importance

@harshamv The original description of this issue should now give you this information. If not, please feel free to open a new discussion or issue. We're happy to help.

**Comment by jasonkuhrt** [comment]
https://www.prisma.io/docs/concepts/components/prisma-schema/relations/referential-integrity

**Comment by 2color** [comment]
I would also add another critical point when using Prisma with PlanetScale:

**When `referentialIntegrity` is set to `prisma`, foreign keys are not created which means that no index is created for the relation scalar**.

This leads to full table scans when the relations for a query are also fetched as explained in [this blog post](https://briananglin.me/posts/spending-5k-to-learn-how-database-indexes-work/)

To avoid full table scans, it's recommended to add an index on the relation scalars, for example:

```prisma
model Post {
  id       Int       @id @default(autoincrement())
  title    String
  excerpt  String
  content  String?
  views    Int       @default(0)
  likes    Int       @default(0)
  comments Comment[]
}

model Comment {
  id      Int    @id @default(autoincrement())
  comment String
  postId  Int
  post    Post   @relation(fields: [postId], references: [id], onDelete: Cascade)

  @@index([postId])
}
```

The index defined by `  @@index([postId])` will ensure that relations are fetched using an index.

**Comment by PhilBookst** [comment]
Should we modify the connection string when using planetscale? Since there a much higher connection limits (10k for the pro plan) the default of `num_physical_cpus * 2 + 1` cannot be applied here, right? 

Another question I have is: How are explicit m-n relations handled with planetscale?

let's say i have this schema:

```
model Profile {
  id              String?      @db.VarChar(255) @unique
  username        String     @db.VarChar(255) @id 
  profile_pic_url String?     @db.VarChar(255)

  SearchProfiles SearchProfiles[]
  
  @@map("profile")
}

model Search {
  term      String           @db.VarChar(255) @id 
  profiles  SearchProfiles[]

  @@map("search")
}

model SearchProfiles {
  order    Int
  term     String @db.VarChar(255)
  username String @db.VarChar(255)

  search  Search  @relation(fields: [term], references: [term])
  profile Profile @relation(fields: [username], references: [username])

  @@id([term, username]) // <- multicolumn index or rather @@id([term]) or @@id([username]) ?
}
```
with my read query looking like this:

```
export default async function read(term) {
  const search = await prisma.search.findUnique({
    where: {
      term,
    },
    include: {  
      profiles: {  // <- SearchProfiles
        orderBy: {
          order: "asc",
        },
        include: { // <- Profile
          profile: {
            select: {
              profile_pic_url: true,
            },
          },
        },
      },
    },
  });

  return search;
}
```

Is a multi colum index the right choice when using include statements in the query?

Thanks and keep up the great work guys!

**Comment by bogdansoare** [comment]
@2color thank you for the clarification with using indexes.

I have a question on how you should define indexes when you have multiple relationships on the same model.

For example I have this Comment Model:
```prisma
model Comment {
  id        String   @id @default(uuid())
  content   String   @db.VarChar(500)
  createdAt DateTime @default(now())
  post      Post     @relation(fields: [postId], references: [id], onDelete: Cascade)
  postId    String
  author    User     @relation(fields: [authorId], references: [id], onDelete: Cascade)
  authorId  String
}
```

How should I specify the indexes 
```prisma
@@index([authorId])
@@index([postId])
```

or 
```prisma
@@index([authorId, postId])
```
or both
```prisma
@@index([authorId])
@@index([postId])
@@index([authorId, postId])
```
?

**Comment by 2color** [comment]
Hey @bogdansoare,

Deciding which indices you should define largely depends on your query patterns. 

Assuming you want to optimise relation lookups, I would recommend defining two single column indices:
```prisma
@@index([authorId])
@@index([postId])
```

The compound index example you shared (`@@index([authorId, postId])`) would only be useful if you are querying the `Comment` table and filtering based on both `authorId` and `postId`.

It's also worth noting that adding many indices has a cost – the index will need to be updated every time you write to the table.

Check out [this article](https://user3141592.medium.com/single-vs-composite-indexes-in-relational-databases-58d0eb045cbe) which covers how indices work in more detail.

**Comment by u007** [comment]
hi,
tested connection working, db migrate works,
but npx prisma db seed failed:

```
  10 
→ 11 const user1 = await prisma.user.create(
  Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(Server(ServerError { code: 1366, message: "target: unigate.-.primary: vttablet: (errno 1366) (sqlstate HY000) (CallerID: unsecure_grpc_client): Sql: \"insert into `User`(id, createdAt, updatedAt, email, `password`, firstname, lastname, role) values (:v1, :v2, :v3, :v4, :v5, :v6, :v7, :v8)\", BindVars: {}", state: "HY000" })) })
    at cb (/home/james/www/unigateBackend/node_modules/@prisma/client/runtime/index.js:38679:17) {
  clientVersion: '3.5.0'
}
```

**Comment by sudokud** [comment]
hello, thanks for the support, any one has an idea on how to automate updating the branch database url in the .env file when we create a new database branch on planetScale?

**Comment by janpio** [comment]
The `.env` file is just a convenient way to put values into your environment variables, that Prisma supports out of the box. If there already is a value in the environment variable, it is not overwritten. So you can just set that env var in a script (e.g. the one you use to create the branch, something like `export DATABASE_URL="foo"` or similar depending on your OS).

**Comment by rvlewerissa** [comment]
Hey @2color,

Do you think we need to add `@@index()` for an implicit many-to-many relationships?

**Comment by 2color** [comment]
Hey @rvlewerissa,

Based on the same logic [as the comment above](https://github.com/prisma/prisma/issues/7292#issuecomment-963118192), it would be recommended to add two indices for each of the foreign keys in the relation table.

The challenge is that [Implicit _many-to-many_ relations](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/many-to-many-relations#implicit-many-to-many-relations) are managed by Prisma and their relation table is **not** surfaced in the Prisma schema. 

That means that you would have to edit the generated SQL and add the index there.

**Comment by flybayer** [comment]
## Question for Prisma

Is `prisma migrate deploy` not compatible with planetscale? (I can't get it to work). If not, will it be? Feels so weird to have to use `migrate dev` and two planetscale branches to run migrations.

## Question for Everyone

How are y'all automating production migrations and syncing them with new deployments? Any code you can share?

**Comment by rvlewerissa** [comment]
@flybayer Hi, maybe you want to check [this helper script](https://github.com/jonico/pscale-cli-helper-scripts) (I haven't tried it myself tho).

**Comment by remorses** [comment]
@flybayer i personally never use `prisma migrate`, instead i directly do `prisma db push` on the non production Planetscale branch and then open a deploy request on main in the Planetscale dashboard, here i check that everything is fine and merge

**Comment by flybayer** [comment]
@remorses I did try that and it works, but I don't like the fact that `db push` skips the migrations table and provides no guarantees around which migrations are applied or not. But maybe it fits better with the planetscale workflow? 🤔

**Comment by tomhoule** [comment]
The reasoning we have in the Migrate team is that `db push` is the right way. Migrations have several important reasons to exist, but most of them don't apply in the Planetscale branching model:

- Determinism: what is run is exactly what you wrote and reviewed, and it's the same in development and production. It doesn't apply on PlanetScale because _planetscale will not run your migrations when you merge your dev branch_, it will generate its own by diffing the schemas from both branches and figuring out its own migration path.
- Audit trail: see what was applied when. Because the schema changes are actually applied on branch merge and not when the migrations are run (and marked as run in the migrations table), the migrations table doesn't have much value on PlanetScale.
- Control: you can't customize what `db push` does, but you are free to do whatever you want in a migration. Again, this doesn't apply on PlanetScale because when the branch is merged, it's not your migration that will be run but something generated by PlanetScale.

**Comment by flybayer** [comment]
@tomhoule cool, thank you. That makes sense. And in that case, we don't need PlanetScale's "automatically copy prisma migrations" feature either.

**Comment by Vandivier** [comment]
additional info and a sample app at https://twitter.com/taylor_atx/status/1469332609847242753

**Comment by jde** [comment]
Hey all. Thank you so much for making prisma work with planetscale, considering its lack of foreign keys. 

I'm struggling to make the referentialIntegrity flag work. I've included all the details here: https://stackoverflow.com/questions/70356527/why-isnt-referentialintegrity-prisma-preventing-foreign-keys-in-migrations

TLDR: I'm using prisma 3.6 with the referentialIntegrity Preview Feature. My schema file begins with:

```
datasource db {
  provider             = "mysql"
  url                  = env("DATABASE_URL")
  shadowDatabaseUrl    = env("SHADOW_DATABASE_URL")
  referentialIntegrity = "prisma"
}

generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["referentialIntegrity"]
}

```

Yet, I'm still seeing this:

```
$ npx prisma migrate dev
...

Database error code: 1317

Database error:
Foreign keys cannot be created on this database. Learn more how to handle this: https://pris.ly/d/migrate-no-foreign-keys
```

Am I missing a step or a concept? Why is prisma trying to create foreign keys with those flags set?

Thanks in advance for any advice!

**Comment by jde** [comment]
Hey all. I've been using prisma with planetscale for a week now and am hitting what I believe to be either a bug or an incompatibility with standard MySQL in the planetscale db. 

When updating, prisma creates many queries based on the relationships in the schema. Here's the relevant parts of the schema I'm using:

```
model Trip {
    id          Int                 @id @default(autoincrement())

    organizer   User                @relation(fields: [organizerId], references: [id])
    organizerId String
    services    ServicesOnTrips[]

    @@map("rd_trip")
}

model Service {
    id          Int         @id @default(autoincrement())
...
}
model ServicesOnTrips {
    trip      Trip    @relation(fields: [tripId], references: [id])
    tripId    Int
    service   Service @relation(fields: [serviceId], references: [id])
    serviceId Int

   @@map("rd_services_on_trips")
}

model User {
  id            String    @id @default(cuid())
  contact       String    @default("SMS and Email")

  trips         Trip[]

  @@map("auth_user")
}

```

As you can see, Trips have an organizer who's a user. They also are related to services through the many to many mapping model ServicesOnTrips. Pretty basic relationships.

When I try to update a user, hoewever, I'm getting an error:

```
    await prisma.user.update({
        where: {
            id: s.user.id,
        },
        data: {
            contact: contact
        }
    })

outputs:
PrismaClientUnknownRequestError: 
Invalid `prisma.user.update()` invocation:

  Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(Server(ServerError { code: 1105, message: "symbol rolldeep.rd_trip.id not found", state: "HY000" })) })
    at cb (/Users/jde/dev/github/jde/rolldeep/node_modules/@prisma/client/runtime/index.js:38687:17)
```

Note that the error is `symbol rolldeep.rd_trip.id not found`. 

I can confirm that in the database `rolldeep`, table `rd_trip`, there indeed is an id field. 

After outputting the query log from prisma, I can see that this symbol is being used in two different queries:

```
Query: SELECT `rolldeep`.`rd_trip`.`id`, `rolldeep`.`rd_services_on_trips`.`serviceId` FROM `rolldeep`.`rd_services_on_trips` WHERE 1=0

and

SELECT `rolldeep`.`rd_trip`.`id`, `rolldeep`.`rd_trip`.`organizerId` FROM `rolldeep`.`rd_trip` WHERE (1=1 AND `rolldeep`.`rd_trip`.`organizerId` IN (?))
```

These relations aren't relevant to my update statement.

Is this an issue with prisma writing malformed sql? Or is this a planetscale specific issue?

At least one other user is experiencing this: https://github.com/prisma/prisma/discussions/10755

**Comment by janpio** [comment]
I think that sounds similar to a few of the other issues we recently got about referentialIntegrity and PlanetScale, check some out at https://github.com/prisma/prisma/issues?q=is%3Aopen+label%3A%22topic%3A+database-provider%2Fplanetscale%22+sort%3Aupdated-desc. In any case, please move your comment into its own issue @jde so we can investigate properly. Thanks.

**Comment by janpio** [comment]
> Lesson: prisma will not recognize differences in migrations _caused by these kinds of config changes._ You'll have to delete the effected migrations and recreate them.

@jde I created an issue about this so we can try to improve the situation: https://github.com/prisma/prisma/issues/10843 Not highest priority of course, but sooner or later.

**Comment by janpio** [comment]
I updated the original post:

- PlanetScale made a change that newly created databases' `main` branch initially is not marked as the production branch, which means you can make schema changes to it by default. This removes the necessity to first create a branch, then use `db push` to make the schema changes and then merging that back via a deploy request. I moved the mention of deploy requests etc to the bottom of the list of things to know.

**Comment by jde** [comment]
Thanks for upgrading this to an issue @janpio!

**Comment by focux** [comment]
@janpio I'm a bit curious about why you didn't incorporate the recommendation about preferring the use of `db push` over `migrate dev/deploy` on the original post? Is it because you aren't still sure about it?

**Comment by janpio** [comment]
Because I am technically on vacation 😆 `db push` is the right thing to use with PlanetScale.

**Comment by jonahallibone** [comment]
I believe I am having trouble with this same issue in my schema. I have a Menu model and a Section model and `1:M` from Menu -> Section


```
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["referentialIntegrity"]
}

datasource db {
  provider             = "mysql"
  url                  = env("DATABASE_URL")
  referentialIntegrity = "prisma"
}

model Menu {
  id        String     @id @default(cuid())
  title     String
  userId    String
  language  String
  currency  String
  createdAt DateTime   @default(now())
  updatedAt DateTime   @updatedAt
  theme     String?
  published Boolean    @default(false)
  user      User       @relation(fields: [userId], references: [id], onDelete: Cascade)
  sections  Section[]
  items     MenuItem[]

  @@unique(fields: [id, userId], name: "menuId_userId_unique")
}

model Section {
  id        String                @id @default(cuid())
  menuId    String
  hidden    Boolean               @default(false)
  deleted   Boolean               @default(false)
  title     String
  order     Int                   @default(0)
  createdAt DateTime              @default(now())
  updatedAt DateTime              @updatedAt
  menu      Menu                  @relation(fields: [menuId], references: [id], onDelete: Cascade, onUpdate: NoAction)
  addons    SectionAddOn[]
  items     MenuItemsOnSections[]

  @@index([menuId])
}
```

When I try to perform an update/delete on a section
```
const menu = await prisma.menu.update({
  where: {
    menuId_userId_unique: {
      id,
      userId: session.user.id,
    },
  },
  data: {
    sections: {
      delete: {
        id: sectionId,
      },
    },
  },
  include: {
    sections: {
      include: {
        items: {
          include: {
            menuItem: true,
          },
        },
      },
    },
  },
});
```

I get them following error:
```
Invalid `prisma.menu.update()` invocation:

Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(Server(ServerError { code: 1105, message: "symbol `example-db`.Section.id not found", state: "HY000" })) })
```

I've tried to follow previous advice by adding `onUpdate: NoAction` to my relations, however, this does not seem to remedy this issue with deletes. Is there something more I should be doing to make this delete behavior work with PlanetScale?

**Comment by 2color** [comment]
Hey folks, 
We just published a video on Prisma & PlanetScale best practices. It covers a lot of the knowledge encapsulated in this issue:

[![Prisma Planetscale Best Practices Guide](https://user-images.githubusercontent.com/1992255/148994499-4557004f-1471-4e19-850e-5946a76ccc15.png)](https://www.youtube.com/watch?v=iaHt5_hg44c)

**Comment by janpio** [comment]
Our mighty documentation team just published a much better version of all the content of this issue in our documentation: 

[![image](https://user-images.githubusercontent.com/183673/151360070-5290a531-daa5-47b3-9129-45d8e5825e91.png)
](https://www.prisma.io/docs/guides/database/using-prisma-with-planetscale
)
https://www.prisma.io/docs/guides/database/using-prisma-with-planetscale

Please take a look and follow the instructions in there from now on.
If anything does not work on PlanetScale with Prisma, please continue opening issues so we can investigate and fix. You can find existing issues with this link: https://github.com/prisma/prisma/issues?q=is%3Aopen+label%3A%22topic%3A+database-provider%2Fplanetscale%22+sort%3Aupdated-desc

Thanks 👋

**Comment by huv1k** [comment]
I just wonder is `referentialIntegrity` going to be moved out of preview features any time soon?

**Comment by janpio** [comment]
Yes, working on that right now. There are a few bugs left in the emulation unfortunately, so this will not happen in the next release(s), but we definitely want to make this a normal part of Prisma @huv1k.

**Comment by janpio** [comment]
@huv1k It happened, with 4.7.0 there is no more `referentialIntegrity` preview feature any more but this is generally available. Here is our documentation: https://www.prisma.io/docs/concepts/components/prisma-schema/relations/relation-mode

Note the recent addition of validation warnings for indexes: https://www.prisma.io/docs/concepts/components/prisma-schema/relations/relation-mode#indexes This is quite important to make sure your database and queries are fast, and your PlanetScale bill does not explode.

**Comment by end0cr1ne** [comment]
@janpio is it common to get a really high number of P1017: Server has closed the connection errors when using Prisma with planetscale?
I've tried adding: connect_timeout=300 to the connection string and ensuring that I have no other connections to the DB, but nothing seems to work.
Any suggestions??

**Comment by janpio** [comment]
No, definitely not common. Best open a new discussion or issue and provide more information.
